p2p/ipld/read.go

```diff
@@ -3,16 +3,215 @@ package ipld
 import (
 	"context"
 	"errors"
+	"fmt"
 	"math"
 
 	"github.com/ipfs/go-cid"
 	coreiface "github.com/ipfs/interface-go-ipfs-core"
 	"github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/lazyledger/lazyledger-core/p2p/ipld/plugin/nodes"
+	"github.com/lazyledger/lazyledger-core/types"
+	"github.com/lazyledger/rsmt2d"
 )
 
-// /////////////////////////////////////
+// //////////////////////////////////////
+//	Retrieve Block Data
+// ////////////////////////////////////
+
+const baseErrorMsg = "failure to retrieve block data:"
+
+var ErrEncounteredTooManyErrors = fmt.Errorf("%s %s", baseErrorMsg, "encountered too many errors")
+var ErrTimeout = fmt.Errorf("%s %s", baseErrorMsg, "timeout")
+
+// RetrieveBlockData asynchronously fetches block data until the underlying extended
+// data square can be restored and returned.
+func RetrieveBlockData(
+	ctx context.Context,
+	dah *types.DataAvailabilityHeader,
+	api coreiface.CoreAPI,
+	codec rsmt2d.CodecType,
+) (types.Data, error) {
+	edsWidth := uint32(len(dah.ColumnRoots))
+	originalSquareWidth := edsWidth / 2
+
+	// keep track of leaves using thread safe counter
+	sc := newshareCounter(ctx, edsWidth)
+
+	// convert the row and col roots into Cids
+	rowRoots := dah.RowsRoots.Bytes()
+	rowCids, err := rootsToCids(rowRoots)
+	if err != nil {
+		return types.Data{}, err
+	}
+
+	colRoots := dah.ColumnRoots.Bytes()
+	colCids, err := rootsToCids(colRoots)
+	if err != nil {
+		return types.Data{}, err
+	}
+
+	// async attempt to fetch each leaf
+	for axisIdx := uint32(0); axisIdx < uint32(len(rowCids)); axisIdx++ {
+		for idx := uint32(0); idx < edsWidth; idx++ {
+			go sc.retrieveShare(colCids[axisIdx], false, axisIdx, idx, api)
+			go sc.retrieveShare(rowCids[axisIdx], true, axisIdx, idx, api)
+		}
+	}
+
+	// wait until enough data has been collected, too many errors encountered,
+	// or the timeout is reached
+	err = sc.wait()
+
+	if err != nil {
+		return types.Data{}, err
+	}
+
+	// flatten the square
+	flattened := sc.flatten()
+
+	tree := NewErasuredNamespacedMerkleTree(uint64(originalSquareWidth))
+
+	// repair the square
+	eds, err := rsmt2d.RepairExtendedDataSquare(rowRoots, colRoots, flattened, codec, tree.Constructor)
+	if err != nil {
+		return types.Data{}, err
+	}
+
+	blockData, err := types.DataFromSquare(eds)
+	if err != nil {
+		return types.Data{}, err
+	}
+
```

Comment:
before we return the block data, do we want to add the data to the local IPFS?

Reply:
Doesn't getting the data automatically add it to your local IPFS data store as well? That's what I was assuming at least 🤔

Reply:
I should have clarified, the repaired data is not put in the local IPFS store.

Reply:
now that we're only going to get the original data, I'm guessing that we shouldn't bother with recomputing the extended data square and putting that to IPFS?

Reply:
Oh, good point then. That data is never downloaded via IPFS and hence wouldn't end up in the local store. We could (ideally in a non-blocking way) put all nodes into the local store via some NodeVisitor if everything checked out.

Reply:
I would assume IPFS is clever enough to not touch the local store if the data did not change. If that is true we could simply call the portion of [`PutBlock`](https://github.com/lazyledger/lazyledger-core/blob/2b2ce8aed3d52945dc34292846fbb39e16e44b3b/types/block.go#L298) that doesn't recompute the EDS but just puts all nodes.

The question here is, what the default should be. If a peer requests enough data to reconstruct the whole block, should it per default also serve the whole EDS, or, should it (per default) only serve the parts it also requested from the network itself. The former would more quickly take some pressure from the block proposer during consensus for instance. cc @adlerjohn 

IMO, this deserves a separate issue for discussion. We can keep it as is for now (default = only serve data requested).

Reply:
Resolving as this is now tracked here: https://github.com/lazyledger/lazyledger-core/issues/279